TST: Skip scipy tests if scipy isn't installed

greglucas opened the pull request:
Scipy isn't a required dependency, so skip the tests that require scipy if scipy isn't found.

`pytest.importorskip` in the module-level imports caused flake8 warnings, so I decided to just try/except and skip instead which passes flake8 without requiring a bunch of extra ignores.

closes #2275

rcomer commented on lib/cartopy/tests/test_img_transform.py:
```diff
@@ -7,7 +7,13 @@
+try:
+    import scipy.spatial
+except ImportError:
+    pytest.skip("scipy is required for image transforms", allow_module_level=True)
```
In here, everything except `test_regridding_with_invalid_extent` will pass providing you have at least one of `scipy` or `pykdtree` installed.  Is it worth modifying this to try both imports?  Or is it too unlikely that a developer would have `pykdtree` but not `scipy`?

I am also unclear how to add appropriate skips to the parametrization of `test_regridding_with_invalid_extent` if we did that...

greglucas replied:
Sure, it is simple enough to do the nested try/except and put the skip only at the nested version. I also added a skip to the parameterization if scipy can't be imported there.

greglucas commented on lib/cartopy/tests/mpl/test_contour.py:
```diff
@@ -8,8 +8,13 @@
+    from scipy.interpolate import NearestNDInterpolator
+    from scipy.signal import convolve2d
+except ImportError:
+    pytest.skip("scipy is required for contouring", allow_module_level=True)
```
Good point. Moved that down into the test.

rcomer reviewed (approved):
If I run without `scipy` or `pykdtree`, I get failures in `test_examples.py` and `test_imges.py`.  That could arguably be out of scope of this PR though.

greglucas commented:
> If I run without scipy or pykdtree, I get failures in test_examples.py and test_imges.py. That could arguably be out of scope of this PR though.

Good catch, I didn't realize I had pykdtree hanging around still. In fact, I get even more `ERRORS` due to image dependency imports in other modules like the `ows` tests. This is adding a lot of conditional logic for the skips in many places at this point considering no one else has complained about this before! I'm not sure how valuable this is...

dopplershift commented:
I think it's definitely helpful given that neither of these are considered required, so IMO we should be able to run the test suite successfully without them.

dopplershift commented:
I guess the alternative is to add one/both to the "test" optional dependencies.

greglucas commented:
Alright, I refactored it a bit and tested this with the various combinations of pykdtree/scipy available/not available so that there are no errors.

There are a few problems that were slightly annoying and made this maybe a little more verbose than ideal:
1. We import our img_transform in other modules, for example to get `_OWS_LIB_AVAILABLE` we go through that entire module that has an import of img_transform, which requires us to skip above that import statement. This also means we can't set `pytestmark = requires_pykdtree_scipy` at the top of the module because that occurs after the module imports.
2. We have optional dependencies on two imports scipy or pykdtree, so `pytest.importorskip` I don't think will work in that situation. So I added a `_HAS_PYKDTREE_SCIPY` variable that we then use to just apply a `pytest.skip()` early on in the test module imports.

QuLogic commented:
Seems to still be conflicts now?

QuLogic commented on lib/cartopy/tests/test_img_transform.py:
```diff
@@ -7,7 +7,17 @@
 import pytest
-import scipy.spatial
+
+from cartopy.tests.conftest import (
+    _HAS_PYKDTREE,
+    _HAS_SCIPY,
+    requires_pykdtree,
+    requires_scipy,
+)
+
+
+if not (_HAS_PYKDTREE or _HAS_SCIPY):
+    pytest.skip("pykdtree or scipy are required", allow_module_level=True)
```
```suggestion
from cartopy.tests.conftest import (
    _HAS_PYKDTREE_OR_SCIPY,
    requires_pykdtree,
    requires_scipy,
)


if not _HAS_PYKDTREE_OR_SCIPY:
    pytest.skip("pykdtree or scipy are required", allow_module_level=True)
```

QuLogic commented on lib/cartopy/tests/test_vector_transform.py:
```diff
@@ -6,6 +6,14 @@
 
 import numpy as np
 from numpy.testing import assert_array_almost_equal, assert_array_equal
+import pytest
+
+
+try:
+    import scipy  # noqa: F401
+except ImportError:
+    pytest.skip("scipy is required for vector transforms", allow_module_level=True)
+
 
 import cartopy.crs as ccrs
 import cartopy.vector_transform as vec_trans
```
```suggestion
from cartopy.tests.conftest import requires_scipy

import cartopy.crs as ccrs
import cartopy.vector_transform as vec_trans


pytestmark = requires_scipy
```

greglucas replied:
This one unfortunately doesn't work because `import cartopy.vector_transform` raises an import error before the pytestmark takes effect.

rcomer reviewed (approved):
Sorry for previously messed up review - I started adding comments a couple of weeks ago but then PRed that change instead.  Then forgot I still had the comment here when I approved it.  Can’t seem to delete the comment now either.

Anyway, I think this is good to go.

QuLogic commented:
Needs a rebase.
